File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/f77-lang-matvect/mv_mult_slave_sschd.f
c     
c*********************************************************************
c
c	C-DAC Tech Workshop : hyPACK-2013
c                 October 15-18, 2013
c
c   Example 17		: mv_mult_slave_sschd.f
c
c   Objective           : Matrix_Vector Multiplication
c                         (Self_scheduling algorithm slave Program)
c
c   Input               : A matrix row from master
c
c   Output              : Computed result to master
c
c
c   Necessary Condition : Number of processors should be greater than
c                         2 and less than or equal to 8
c
c   Created             : August-2013
c
c   E-mail              : hpcfte@cdac.in     
c
c*********************************************************************
c
      
      program worker

      include 'mpif.h'

      integer MAX_ROWS, MAX_COLS 
      parameter (MAX_ROWS = 1000, MAX_COLS = 1000)

      double precision Vector(MAX_COLS)
      double precision MyBuffer(MAX_COLS), ans
c
      integer MyRank, Root, Numprocs, ierr, status(MPI_STATUS_SIZE)
      integer i, Destination_tag, VectorSize, ValidInput

      call MPI_INIT( ierr )
      call MPI_COMM_RANK( MPI_COMM_WORLD, MyRank, ierr )
      call MPI_COMM_SIZE( MPI_COMM_WORLD, Numprocs, ierr )

      Root   = 0
      ValidInput = 1

      if(Numprocs .lt. 2) then
	 goto 400
      endif

      call MPI_Barrier(MPI_COMM_WORLD, ierr)

      call MPI_Bcast(ValidInput, 1, MPI_INTEGER, Root, 
     $          MPI_COMM_WORLD, ierr)

       if(ValidInput .eq. 0) then
	 goto 400
       endif


C 	Slavereceives VacorSize and Vector

 	 call MPI_BCAST(VectorSize, 1, MPI_INTEGER, Root,
     $        MPI_COMM_WORLD, ierr)

 	 call MPI_BCAST(Vector,VectorSize,MPI_DOUBLE_PRECISION,Root,
     $        MPI_COMM_WORLD, ierr)

c  	Slave receive B, then compute dot product until done message.

90      call MPI_RECV(MyBuffer,VectorSize,MPI_DOUBLE_PRECISION,Root,
     $        MPI_ANY_TAG, MPI_COMM_WORLD, status, ierr)

            flag = status(MPI_TAG)
         if (flag .eq. 0) then 
             go to 200
         else
	      Destination_tag = flag
	      ans = 0.0
              do 100 i = 1, VectorSize
                ans = ans + MyBuffer(i)*Vector(i)
100	      continue               

              call MPI_SEND(ans, 1, MPI_DOUBLE_PRECISION, Root,
     $                 Destination_tag, MPI_COMM_WORLD, ierr)
              go to 90
          endif
200	continue

        write(6,*) 'slave  successful  ', MyRank, ' is done'
400	call MPI_FINALIZE(ierr)

        stop
        end
              

c	********************************************************************

File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/c-lang-matvect/mv_mult_slave_sschd.c
/*
********************************************************************	

		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

   Example 5.5	       : mv_mult_slave_sschd.c

   Objective           : Matrix_Vector Multiplication
                         (Self_scheduling algorithm worker Program)

   Input               : A matrix row from master 

   Output              : Computed result to master 

   Necessary Condition : Number of processes should be greater than 
                         2 and less than or equal to 8

   Created             : August-2013

   E-mail              : hpcfte@cdac.in     

******************************************************************/




#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

main(int argc, char** argv) 
{

	int		  Numprocs, MyRank;
	int		  Source, Destination;
	int		  Source_tag, Destination_tag;
	float		  *MyBuffer, *Vector, Sum;
	int		  VectorSize;
  	int 		  index, flag, ValidInput = 1;
	int	     Root = 0;
	int		  MatrixFileStatus = 1, VectorFileStatus = 1;
  	MPI_Status status;     

   /* ........MPI Initialisation .......*/
  	MPI_Init(&argc, &argv); 
  	MPI_Comm_rank(MPI_COMM_WORLD, &MyRank);
  	MPI_Comm_size(MPI_COMM_WORLD, &Numprocs);

	if(Numprocs < 2) {
	   MPI_Finalize();
	   exit(0);
	}  

   MPI_Bcast (&MatrixFileStatus, 1, MPI_INT, Root, MPI_COMM_WORLD);
	if(MatrixFileStatus == 0) {
	   MPI_Finalize();
           exit(-1);
	}

   MPI_Bcast (&VectorFileStatus, 1, MPI_INT, Root, MPI_COMM_WORLD);
	if(VectorFileStatus == 0) {
	   MPI_Finalize();
           exit(-1);
	}

        MPI_Bcast(&ValidInput, 1, MPI_INT, Root, MPI_COMM_WORLD);

	if(ValidInput==0){
           MPI_Finalize();
	   exit(0);
	}

	MPI_Bcast(&VectorSize, 1, MPI_INT, 0, MPI_COMM_WORLD); 

        Vector = (float *)malloc(VectorSize*sizeof(float));   
	MPI_Bcast (Vector, VectorSize, MPI_FLOAT, 0, MPI_COMM_WORLD); 

        MyBuffer = (float *)malloc(VectorSize*sizeof(float));   

        for(;;) {
	   MPI_Recv(MyBuffer, VectorSize, MPI_FLOAT, Root, MPI_ANY_TAG, 
					 MPI_COMM_WORLD, &status);
           flag = status.MPI_TAG - 1;

	   if(flag == -1) 
	      break;

	   Destination = 0;
	   Destination_tag = flag;

	   Sum = 0.0;
	   for(index=0; index< VectorSize; index++) 
	       Sum += (MyBuffer[index] * Vector[index]);
	   MPI_Send(&Sum, 1, MPI_FLOAT, 0, Destination_tag,MPI_COMM_WORLD);
        }

	free(MyBuffer);
	free(Vector);

        MPI_Finalize();
}



